Fill in Python SQA Documentation

cticenhour opened the pull request:
@permcody This is a good start at filling in SQA documentation for MOOSE Tools so that we can close the labway issue. I plan to finish filling all of this in next week, but I wanted to make a PR to let you review and comment on what is here (if you have time - I can ping you again when it's done if not). 

Refs #15968

cticenhour committed:
Fix python sqa doc typos (#15968)

cticenhour committed:
Switch app python SQA docs to the base template

cticenhour committed:
Fill in python software test plan and fix typos in framework version

cticenhour committed:
Switch from app to project in python sqa documentation

cticenhour committed:
Fixup framework scs and point python sqa to right location

cticenhour committed:
Fill in most of python sqa failure analysis report

Need to come back to failure analysis section concerning types of failure output

cticenhour committed:
Fill in most of python sqa requirements traceability matrix

Need to fill in system scope, system purpose, and assumptions and dependencies, which also applies to other documents

cticenhour committed:
Centralize all minimum system requirements markdown and add python as a tracked minimum version

cticenhour committed:
Fill out most of the python sqa software requirement specification

The system purpose, system scope, assumptions and dependencies, and system modes still need to be filled in

cticenhour committed:
Remove category and app template parameters for python sqa docs where not necessary

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1310727/) on e96ba83 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23217/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1310763/) on e96ba83 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>9f4eda</th>
      <th colspan=4>#23217 e96ba8</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.61%</td>
      <td>84.61%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>81968</td>
      <td>81964</td>
      <td>-4</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14906</td>
      <td>14910</td>
      <td>+4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23217/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23217/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23217/coverage/xfem)

</details>



This comment will be updated on new commits.

cticenhour committed:
Fill in python software library list sqa documentation

cticenhour commented on framework/doc/content/sqa/minimum_requirements.md:
```diff
@@ -6,7 +6,10 @@ In general, the following is required for MOOSE-based development:
-- Many flavors of modern Linux and the two most recent versions of MacOS
+- Memory: 8 GBs of RAM for optimized compilation (16 GBs for debug compilation), 2 GB per core execution
+- Processor: 64-bit x86 or ARM64 (specifically, Apple Silicon)
+- Disk: 100GB
```
Would welcome some comment here on minimum disk usage - we had 30GB in one place and 100GB in another....

cticenhour replied:
I had the same thought when I saw the descrepancy.....there's no way in heck this is the "minimum"....I kept it in to prompt a discussion.

cticenhour replied:
After building the main three methods (`opt`,`devel`, and `dbg`), the size of the MOOSE directory was about 21GB, so I think I will stick with 30GB here. Updated in the most recent commit.

GiudGiud replied:
did you build combined in all 3? It's excessive to me

cticenhour replied:
I don't agree that 30 is excessive. By building all three combined executables, @friedmud and I wanted to create a worst case scenario where one might want multiple copies of the complete executable in place with comfortable space for simulations, analysis, and debugging. Derek suggested adding 10GB to this, while I felt 9 (to get to a round number) was acceptable. 30GB happens to also be the min drive space we've suggested to people for years, with 100GB being the anomaly, so I'm comfortable with this.

cticenhour committed:
Add system purpose and system scope documentation for python sqa

This fills in gaps in the SRS, RTM, and SDD

cticenhour committed:
Add assumptions and dependencies to python sqa SDD, RTM, and SRS

cticenhour committed:
Fill in python sqa software design description

cticenhour committed:
Finish python sqa system requirements specification

cticenhour committed:
MOOSE tools test spec fixes

- Switch to 'The system shall' in MOOSE tools requirements
- Clean up test file syntax
- Add failure analysis collection parameters to relevant TestHarness tests

cticenhour committed:
Finish up python sqa failure analysis report

cticenhour committed:
Remove usage of project parameter in python STP body text

cticenhour commented on python/doc/content/python/sqa/python_srs.md:
```diff
@@ -1 +1,135 @@
+!template item key=human-system-integration
+[!ac](MOOSE) Tools contains command line driven applications which conform to all standard terminal
+behaviors. Specific human system interaction accommodations shall be a function of the end-user's
+terminal. Optional accomodations will be outlined in utility design documents.
```
```suggestion
terminal. Optional accommodations will be outlined in utility design documents.
```

GiudGiud commented on python/doc/content/python/sqa/python_srs.md:
```diff
@@ -1 +1,135 @@
+- +Analysts+: These are users that will run the code and perform various analysis and testing on the
+  simulations they perform. These users may interact with developers of the system requesting new features
+  and reporting bugs found and will typically make heavy use of the pre-built scripts, configuration
+  files, and the MooseDown file format.
```
add a link to moosedown?

cticenhour commented on python/doc/content/python/sqa/python_sll.md:
```diff
@@ -1 +1,269 @@
-!template load file=sqa/app_sll.md.template category=python app=MOOSE Tools
+!template load file=sqa/sll.md.template project=MOOSE Tools
+
+## NumPy
```
"MOOSE Tools" encompasses mooseutils, MooseDocs, TestHarness and the miscellaneous utilities in the `python` dir. We explicitly leave out Peacock and Chigger for the purposes of SQA, because they're truly optional, so VTK isn't necessary for this list.

cticenhour commented on python/TestHarness/tests/tests:
```diff
@@ -3,262 +3,276 @@
     type = PythonUnitTest
     input = test_DuplicateTestNames.py
-    requirement = "TestHarness shall skip tests having the same name contained in the tests file"
+    requirement = "The system shall skip tests having the same name contained in the tests file"
```
Looking at #10424 and #10428, the previous behavior was just a confusing error, as the Scheduler wasn't equipped to handle the case in the first place. Once the more graceful skip occurs, the Parser then catches the duplicate name (and perhaps duplicate params for a "true" duplicate test) and errors out - see [this comment](https://github.com/idaholab/moose/issues/10424#issuecomment-352770091) for an example of the current output when there are multiple matches. 

So, the end result of this isn't necessarily dangerous, as an error will still be thrown by the parser system for the developer to address....it's not a true skip. The requirement wording might be able to be improved though... any ideas? My first one is:

```
The system shall allow the test scheduler to gracefully skip tests having the same name contained within the same tests file for further checks against the test file parser.  
```

cticenhour commented on python/doc/content/python/sqa/python_stp.md:
```diff
@@ -1 +1,164 @@
+test is accepted (pass), otherwise it is rejected (fail).
+
+In addition, test coverage reports will be created for all proposed changes. The overall line
+coverage is required to increase or remain constant (it may not decrease). Additionally, the
```
This is also the wording in the main framework documents, I believe. Perhaps we need to soften this language @permcody? We have had instances where coverage decreased based on reviewer acceptability (even if not ideal), though we have generally had a hard limit at 80% in the past.

permcody replied:
Yeah, This is too strong. We don’t strictly enforce this. We do have a hard floor (80%).

cticenhour replied:
Softened this language and also percolated it over to the framework STP.

cticenhour committed:
Address review comments

cticenhour commented on python/doc/content/python/sqa/python_sdd.md:
```diff
@@ -1 +1,145 @@
+this design goal, [!ac](MOOSE) Tools utilities generally use a particular design pattern, consisting
+of a central core of common functionality extended by code designed for specific tasks. Users needing
+to extend or create new [!ac](MOOSE) Tools utilities may, in many cases, use the main code as Python
+packages and specialize or modify it to provide an implementation meeting his or her needs. The
```
Changed this in a few more places (including where I copied from). Thanks for pointing this out!

cticenhour commented:
@GiudGiud Thanks very much for your input in your earlier review. Any more comments? If not, I think this might be ready to go in.

GiudGiud commented:
Gave it another read, looks great
